Feat/setting vc UI

isGeekCode opened the pull request:
## Features ✨
- SettingViewController 화면구현
- SettingTableView 구현
- SettingTableViewCell 구현
- 버전정보를 가져오는 함수 구현
- 칸별 AccessoryType 분기처리
- Extension: NSAttributedString+ 추가

## Screenshots 📸
<img width="415" alt="스크린샷 2022-12-05 오후 6 24 16" src="https://user-images.githubusercontent.com/76529148/205602966-68069b9f-15dd-4cc6-9db2-e527aa03f9bd.png">

## To Reviewers
브랜치 커밋이 커져서 먼저 머지하려고 합니다.
남은 작업은 아래 순서대로 진행할 예정입니다.
색상은 구역확인을 위해 들어가있습니다.
네비게이션 바 구현 시점에서 삭제예정입니다.

남은 작업
- 네비게이션바
- Coordinator
- 셀의 세부내역

NSAttributedString 익스텐션에 구현한 withLineSpacing()은 자간을 조절하는 함수입니다. 버전정보에 사용하였습니다.ㅌ

화면 실행을 하기위해선 
SceneDelegate에서 직접 rootVC = SettingVC()로 구현해야합니다. 머지를 위해 원복 시켰습니다.

전부 UI작업이라 여기까지 한 브랜치에 작업했는데 혹시 너무 과한가요?

isGeekCode committed:
[ADD] SettingVC init

isGeekCode committed:
[ADD] SettingTableViewCell init

isGeekCode committed:
[CHORE] SettingVC - TableView init

isGeekCode committed:
[CHORE] SettingTableViewCell - UI작업

isGeekCode committed:
[CHORE] SettingVC - TableView Row 6으로 변경

isGeekCode committed:
[ADD] SettingVC - 이미지 아이콘 추가

isGeekCode committed:
[CHORE] SettingVC - UI 수정

isGeekCode committed:
[CHORE] SettingTableViewCell - set UI

isGeekCode committed:
[CHORE] SettingTableVIew - SettingTableViewCell 연결

isGeekCode committed:
[CHORE] 기능별 함수 정리

isGeekCode committed:
[CHORE] SettingTableViewCell - disclousureIndicator 세팅

isGeekCode committed:
[CHORE] VersionInfo Cell 세팅

isGeekCode committed:
[CHORE] SettingTableViewCell - Layout 수정

isGeekCode committed:
[CHORE] 텍스트 자간 세팅하는 함수 추가

isGeekCode committed:
[ADD] Extension - NSAttributedString 추가

isGeekCode committed:
[ADD] 누락된 Project파일 추가

isGeekCode committed:
[RENAME] 변수명 변경 descriptionLabel >>> versionInfoLabel

isGeekCode committed:
[CHORE] 자간 세팅하는 함수 적용

isGeekCode committed:
[CHORE] 주석 수정

isGeekCode committed:
[FIX] Constraints 에러 수정

eung7 commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+    $0.textColor = .gray
+  }
+
+  // MARK: Life Cycle
+  override func awakeFromNib() {
+    super.awakeFromNib()
+  }
+  // MARK: SetSelected
+  override func setSelected(_ selected: Bool, animated: Bool) {
+    super.setSelected(selected, animated: animated)
+  }
```
필요하지 않은 메서드는 지우는 게 좋을 것 같아요!

EnnzoKim reviewed (changes requested):
아래 내용을 확인해주세요!

EnnzoKim commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  // MARK: Life Cycle
+  override func awakeFromNib() {
+    super.awakeFromNib()
+  }
+  // MARK: SetSelected
+  override func setSelected(_ selected: Bool, animated: Bool) {
+    super.setSelected(selected, animated: animated)
+  }
```
사용하지 않은 메서드는 지우는 게 좋을 것 같아요!

isGeekCode replied:
수정했습니다!

EnnzoKim commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+    return appVersion
+  }
+
+  /// 텍스트의 자간을 세팅하는 함수
+  private func setTextSpacing(text: String) -> NSAttributedString {
+    let introduction = NSAttributedString(string: text).withLineSpacing(100)
+    return introduction
+  }
+}
```
```swift
extension String {} 
```
을 통해서 Cell 파일과 분리할 수 있을 것 같아요 !

isGeekCode replied:
Instance member 'setTextSpacing' cannot be used on type 'String'; did you mean to use a value of this type instead?

위 에러 메세지가 발생합니다. String안에 구현할 수 없고, 또 NSAttributedString 에도 구현할 수 가 없다고 하여 유지해야할 듯합니다 혹시 또 다른 방법이 없을까요

WhiteHyun replied:
저도 @eung7 님 의견에 동의합니다! extension으로 빼는 게 더 좋은 거 같아요.
그리고 뺄 때 그대로 함수 붙여넣으시지 마시구 약간만 수정하면 적용되니 한 번 시도해보세요!

isGeekCode replied:
수정하다보니  해당 함수가 단순히 형변환하고 있고 , spacing을 조절하기 위한 함수가 들어가있는 2가지의 기능을 하고 있었습니다. 그래서 형변환 자체는 함수로 빼는 것 자체가 불필요하다고 판단되어 아래와 같이 수정했습니다.

```
let attributeString = NSAttributedString(string: versionString).withLineSpacing(100)
```

EnnzoKim commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+
+  // MARK: - Properties
+  static let identifier = "SettingTableViewCell"
+  private let iconList = ["icon_alert", "icon_noti", "icon_review", "icon_mail", "info_cheer", "icon_version"]
+  private let titleList = ["알림", "공지사항", "리뷰 남기기", "문의하기", "개발자 응원하기", "버전정보"]
```
이 Data들은 SettingViewModel에 배치하는게 좋을 것 같아요!
아직 ViewModel이 만들어지지 않아서 그런건가요?

isGeekCode replied:
네 이건 ViewModel 생성후 분리하도록 하겠습니다!

EnnzoKim commented on PPAK_CVS/Sources/Scenes/Setting/SettingVC.swift:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+  }
```
아직 사용하지 않는 메서드는 지워주셔도 좋을 것 같아요 !

isGeekCode replied:
수정했습니다!

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+import SnapKit
+import Then
+
+class SettingTableViewCell: UITableViewCell {
```
`final` keyword

isGeekCode replied:
수정했습니다!

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+//  Created by hyeonseok on 2022/11/28.
+//
+
+import UIKit
+import SnapKit
+import Then
```
```suggestion
import UIKit

import SnapKit
import Then
```

isGeekCode replied:
아 이부분은 한줄 띄어야하는 거군요

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+
+    // 버전정보 셀
+    let isVersionCell = row == 5
+    [versionInfoLabel].forEach {
+
+      let versionString = isVersionCell ? "v \(versionInfo())" : ""
+      $0.attributedText = setTextSpacing(text: versionString)
+      $0.isHidden = isVersionCell ? false : true
+      $0.textAlignment = .right
+    }
```
불필요한 forEach네용

isGeekCode replied:
아하. 이건 $0을 깔끔하게 생각하시는 듯 하여 이렇게 한건데 그러면 변수명으로 수정하겠습니다

isGeekCode replied:
수정했습니다

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+
+  private func setupConstraints() {
+
+    containerView.snp.makeConstraints {
+      $0.leading.equalToSuperview().inset(30)
+      $0.trailing.equalToSuperview().inset(20)
+      $0.verticalEdges.equalToSuperview().inset(10)
+      $0.centerY.equalToSuperview()
+    }
```
`make`로 변경해주세요

isGeekCode replied:
수정했습니다!

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+  /// 현재 앱의 버전정보를 알려주는 함수
+  private func versionInfo() -> String {
+
+    guard let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String else { return ""}
```
```suggestion
    guard let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String else { return "" }
```

isGeekCode replied:
수정했습니다!

WhiteHyun commented on PPAK_CVS/Sources/Scenes/Setting/SettingVC.swift:
```diff
@@ -0,0 +1,84 @@
+// MARK: - TableView Delegate
+extension SettingViewController: UITableViewDataSource, UITableViewDelegate {
+  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+
```
빈줄 제거

isGeekCode replied:
SettingVC, SettingTableViewCell 모든 빈줄 제거했습니다

WhiteHyun commented on PPAK_CVS/Sources/Scenes/Setting/SettingVC.swift:
```diff
@@ -0,0 +1,84 @@
+class SettingViewController: BaseViewController {
+
+  // MARK: - Properties
+  private lazy var headerBar: UIView = {
+    let view = UIView()
+    return view
+  }()
+
+  private lazy var tableView: UITableView = {
+    let tableView = UITableView()
+    tableView.alwaysBounceVertical = false
+    tableView.dataSource = self
+    tableView.delegate = self
+    tableView.register(SettingTableViewCell.self, forCellReuseIdentifier: SettingTableViewCell.identifier)
+    tableView.backgroundColor = .systemTeal
+    return tableView
+  }()
```
`then` 사용해주세요

isGeekCode replied:
then 세팅했습니다

WhiteHyun commented on PPAK_CVS/Sources/Scenes/Setting/SettingVC.swift:
```diff
@@ -0,0 +1,84 @@
+  override func setupConstraints() {
+    super.setupConstraints()
+
+    headerBar.snp.makeConstraints {
+      $0.top.equalTo(view.safeAreaLayoutGuide)
+      $0.leading.trailing.equalToSuperview()
+      $0.height.equalTo(60)
+    }
+
+    tableView.snp.makeConstraints {
+      $0.top.equalTo(headerBar.snp.bottom).offset(10)
+      $0.leading.trailing.bottom.equalToSuperview()
+    }
```
`make`로 바꿔주세용

isGeekCode replied:
수정했습니다!

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+
+  private func setLayouts() {
+
+    addSubview(containerView)
```
contentView에 넣어야하지 않을까여

isGeekCode replied:
containerView를 contentView로 변수 네이밍을 수정하자는 말씀이실까요

아니면 containerView를 contentView 안에 넣자는 말씀이실까요!!

WhiteHyun replied:
후자입니다

isGeekCode replied:
contentView.addSubView(containerView)로 수정처리했습니다

WhiteHyun commented on PPAK_CVS/Configuration/Extensions/NSAttributedString+.swift:
```diff
@@ -0,0 +1,21 @@
+
+import UIKit
+
+extension NSAttributedString {
+    func withLineSpacing(_ spacing: CGFloat) -> NSAttributedString {
+        let attributedString = NSMutableAttributedString(attributedString: self)
+        let paragraphStyle = NSMutableParagraphStyle()
+        paragraphStyle.lineBreakMode = .byTruncatingTail
+        paragraphStyle.lineSpacing = spacing
+        attributedString.addAttribute(.paragraphStyle,
+                                        value: paragraphStyle,
+                                        range: NSRange(location: 0, length: string.count))
+        return NSAttributedString(attributedString: attributedString)
+    }
+}
```
얘는 어떠한 작업을 맡고있나요?

isGeekCode replied:
네 SettingTableVIewCell  마지막에 구현된 함수인 `setTextSpacing`를 보시면 아래와 같습니다.
```swift
  private func setTextSpacing(text: String) -> NSAttributedString {
    let introduction = NSAttributedString(string: text).withLineSpacing(100)
    return introduction
  }
```
NSAttributedString값으로 파싱된 String 값에 `원하는  글자간격을 세팅하는 함수`입니다.

isGeekCode replied:
주석 추가로 커밋했습니다!

WhiteHyun commented on PPAK_CVS/Sources/Scenes/Setting/SettingVC.swift:
```diff
@@ -0,0 +1,84 @@
+import RxCocoa
+import RxGesture
+
+class SettingViewController: BaseViewController {
```
`final` 사용해주세요

isGeekCode replied:
수정했습니다

WhiteHyun commented on PPAK_CVS/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+  // MARK: setLayout
+  func setUI(_ row: Int) {
+
+    setLayouts()
+    setupConstraints()
```
`init`할 때 layout과 constraint를 설정하는 게 낫지 않을까요?

isGeekCode replied:
@WhiteHyun 님께서 원하시는건 Convinience init()에 넣어서 사용하는 걸까요?
저는 SetUI(row)를 TableView Delegate에서 함께 구현하는게 나을것 같아서 이곳에 해봤습니다.

init()에다가 구현하면 그걸 먼저 하고 TableView Delegate의 Cell.setUI(row)함수가 실행될 수 있을까요

WhiteHyun replied:
제 생각에는 Cell이 **`초기화`** 되었을 때, Cell의 전반적인 View Setting이 되어야하는 게  맞다고 생각해서 드린 말씀이었습니다. 지금은 셀이 적어서 reuse되지 않겠지만, 스크롤 가능한 셀이면 연속으로 layout과 constraint 작업을 중복하여 적용시키기 때문에 차후에 버그가 일어날 가능성이 높은 코드입니다. 수정이 필요해보여요.

isGeekCode replied:
init으로 이동처리했습니다

isGeekCode committed:
[DEL] 사용하지 않는 코드 삭제

isGeekCode committed:
[DEL] SettingVC - 사용하지않는 코드 삭제

isGeekCode committed:
[CHORE] 자간을 세팅하는 함수에 대한 주석 추가

isGeekCode committed:
[CHORE] import 한줄 띄기 수정

isGeekCode committed:
[CHORE] SettingVC 와 SettingTableViewCell 클래스를 final로 변경

isGeekCode committed:
[CHORE] ForEach >> 변수로 원상복구

isGeekCode committed:
[CHORE] snapkit 사용한 메서드의 변수명 수정

- $0 >> make

isGeekCode committed:
[CHORE] 띄어쓰기 수정

isGeekCode committed:
[CHORE] 모든 빈줄 제거

isGeekCode committed:
[CHORE] SettingVC - then 세팅

WhiteHyun commented on PPAK_CVS/Sources/Scenes/Setting/SettingVC.swift:
```diff
@@ -0,0 +1,84 @@
+}
+
+// MARK: - TableView Delegate
+extension SettingViewController: UITableViewDataSource, UITableViewDelegate {
```
MARK 주석 수정해야할 거 같아요. 그리고 사이에 빈 줄 넣어주세요!


https://github.com/StyleShare/swift-style-guide#%EB%B9%88-%EC%A4%84

isGeekCode replied:
수정했습니다!

isGeekCode committed:
[CHORE] MARK: 한줄띄기 수정

isGeekCode committed:
[CHORE]  불필요한 함수 제거

- 해당 함수는 문자열을 속성문자열 (AttributeString)으로 형변환하고 spacing을 조절하는 두가지 기능을 가진 함수이다.
- 단순히 형변환하는 것은 함수로 뻬는 것이 불필요하다고 판단되어 삭제

isGeekCode committed:
[CHORE] SettingTableViewCell - ContainerView를 ContentView에 추가

isGeekCode committed:
[CHORE] init 함수 생성

- 아래 두 함수를 SetUI() -> Init() 으로 이동
- setLayouts
- setupConstraints

isGeekCode committed:
Merge branch 'develop' into feat/settingVC-UI
